### DISCUSSION TITLE
Citation leads to RStudio freeze

### DISCUSSION THREAD

**Discussion by dragonattheend** [quarto-dev/quarto-cli#709] [Q&A] [answered]
When I try to cite something my RStudio session becomes unusable. In particular, if I add a citation like [@example1900] and then remove the brackets (@example1900), I get this "Working..." message on top and cannot click any button. 

![Peek 2022-04-21 17-46](https://user-images.githubusercontent.com/83570780/164565872-e7437005-3d69-4e91-bd8f-5a09668fa5c5.gif)

I run an Archlinux based OS, so not sure if it is the reason for this bug and maybe everything is perfect on Mac and Win.

**Comment by jjallaire** [comment]
One thing that might help is to export a request log to see exactly where things are getting hung up (info on doing that here: https://support.rstudio.com/hc/en-us/articles/200488538-How-To-Export-a-Network-Request-Log-From-the-RStudio-IDE)

**Reply by dragonattheend** [reply]
> One thing that might help is to export a request log to see exactly where things are getting hung up (info on doing that here: https://support.rstudio.com/hc/en-us/articles/200488538-How-To-Export-a-Network-Request-Log-From-the-RStudio-IDE)

Another thing is that the time in log file is different from my system time (Sys.time()).

**Comment by dragonattheend** [comment]
There was a hunspell problem at first, I added .aff, .dic files, but still getting an error. 

`WARNING No memory control group found in /proc/self/cgroup; LOGGED FROM: std::string rstudio::core::system::{anonymous}::getMemoryCgroup() src/cpp/core/system/LinuxResources.cpp:335`

`ERROR CLIENT EXCEPTION (rsession-dat): TypeError: Cannot read property 'inside' of null;|||com/google/gwt/cell/client/AbstractEditableCell.java#41::viewDataMap|||` ... 

`ERROR system error 2 (No such file or directory) [exe: /home/dat/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:819; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:821`

**Comment by jjallaire** [comment, marked as answer]
It looks to me like RStudio is looking for version of Pandoc that doesn't exist  (at `/home/dat/pandoc`). I'm not sure why this would be -- do you have an `RSTUDIO_PANDOC` environment variable set? You might want to try removing that file to see if it clears things up.

Net: this is some sort of misconfiguration or environment conflict which is incredibly hard to debug speculatively.

**Reply by dragonattheend** [reply]
Thank you.
Turns out I had quarto removed, I reinstalled it, and still was getting the error, that the `pandoc` file was missing in `usr/lib/quarto-cli/bin/`, it was in `usr/lib/quarto-cli/bin/tools/`, so I copied it to `usr/lib/quarto-cli/bin/` and it seems to work. Interestingly, I get a different location for `Sys.getenv("RSTUDIO_PANDOC")
[1] "/usr/lib/rstudio/bin/quarto/bin"`. 

Thanks again.

**Comment by jjallaire** [comment]
Makes me think that there could be an Arch packaging issue related to a change we recently made (which was in fact moving `pandoc` from `bin/` to `/bin/tools` !). It's possible that whomever packages RStudio for Arch needs to make a change to reflect this? @timtmok Could you follow up to see if this is the case?

**Comment by rgaiacs** [comment]
@jjallaire I don't think this is a Arch related issue. I updated RStudio and Quarto in my Ubuntu machine

```
$ sudo dpkg -i ~/Downloads/rstudio-2022.02.1-461-amd64.deb                                                                                                                                                                 
(Reading database ... 684683 files and directories currently installed.)                                                                                                                                                                                                                  
Preparing to unpack .../rstudio-2022.02.1-461-amd64.deb ...                                                                                                                                                                                                                               
Unpacking rstudio (2022.02.1+461) over (1.4.1106) ...                                                                                                                                                                                                                                     
Setting up rstudio (2022.02.1+461) ...                                                                                                                                                                                                                                                    
Processing triggers for mailcap (3.69ubuntu1) ...                                                                                                                                                                                                                                         
Processing triggers for desktop-file-utils (0.26-1ubuntu2) ...                                                                                                                                                                                                                            
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...                                                                                                                                                                                                                                 
Processing triggers for hicolor-icon-theme (0.17-2) ...                                                                                                                                                                                                                                   
Processing triggers for shared-mime-info (2.1-1) ...                                                                                                                                                                                                                                      
$ sudo dpkg -i ~/Downloads/quarto-0.9.309-linux-amd64.deb                                                                                                                                                                  
(Reading database ... 686253 files and directories currently installed.)                                                                                                                                                                                                                  
Preparing to unpack .../quarto-0.9.309-linux-amd64.deb ...
Unpacking quarto (0.9.309) over (0.9.238) ...
Setting up quarto (0.9.309) ...
```

and I'm having he same issue. I have

```
2022-04-26T02:58:18.518678Z [rsession-raniere] ERROR system error 2 (No such file or directory) [exe: /opt/quarto/bin/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:819; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:821
```

in RStudio log. I checked the `.deb` from GitHub Releases and pandoc is in `/opt/quarto/bin/tools/` instead of `/opt/quarto/bin/pandoc`.

**Comment by jjallaire** [comment]
Right Pandoc is in `/opt/quarto/bin/tools` but the error message says that RStudio is looking for it at `/opt/quarto/bin/pandoc`. I think the Arch recipe for RStudio needs to be updated to work with the new location of Pandoc in more recent versions of Quarto.

**Reply by trap000d** [reply]
```
2022-04-26T22:41:35.290889Z [rsession-trap000d] ERROR system error 2 (No such file or directory) [exe: /usr/lib/quarto-cli/bin/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/Posi
xChildProcess.cpp:819; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:821
```
Yep, looks like path to quarto (and pandoc correspondingly) is hardcoded somewhere. I'll have a look at 'rstudio-desktop' package and fix it.

**Reply by trap000d** [reply]
Well, I suppose it has to be addressed to RStudio upstream - see rstudio-2022.02.1-461/src/cpp/session/CMakeLists.txt - pandoc location is hardcoded:

```
if(QUARTO_ENABLED)
   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "${RSTUDIO_DEPENDENCIES_QUARTO_DIR}/bin")
elseif(EXISTS "${RSTUDIO_TOOLS_ROOT}/pandoc/${PANDOC_VERSION}")
   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "${RSTUDIO_TOOLS_ROOT}/pandoc/${PANDOC_VERSION}")
else()
   set(RSTUDIO_DEPENDENCIES_PANDOC_DIR "${RSTUDIO_DEPENDENCIES_DIR}/pandoc/${PANDOC_VERSION}")
endif()
```
UPD: It's fixed in RStudio master branch (https://github.com/rstudio/rstudio/pull/10952). So I only need to backport changes.

**Reply by jjallaire** [reply]
Okay, great. This does imply that users need to install new-ish versions of Quarto CLI (but that's essentially where the project is right now -- you pretty much need to be at HEAD). We will be releasing a stable version of Quarto (v1.0) around July.

**Reply by trap000d** [reply]
Well, I've just pushed updates of 'rstudio-desktop' to AUR. Seems quarto works well (though I tested it with 'quarto-cli' package only). Unfortunately, all binary packages still have bug @dragonattheend  reported. 
RStudio developers have missed a place where pandoc is called.

``` patch
diff -ur a/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp b/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp
--- a/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp     2022-04-27 22:28:07.480284489 +1200
+++ b/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp     2022-04-27 20:59:01.396159620 +1200
@@ -188,7 +188,7 @@
    core::system::ProcessResult result;

    error = module_context::runPandoc(
-      FilePath(config.bin_path).completeChildPath("pandoc").getAbsolutePath(),
+      FilePath(config.bin_path).completeChildPath("tools/pandoc").getAbsolutePath(),
       args,
       contents,
       options,
```

**Reply by jjallaire** [reply]
This is actually fixed in the patch release of RStudio 2022.02. Here is the correct fix that should be backported to AUR (note it handles both possible locations for pandoc): https://github.com/rstudio/rstudio/blob/prairie-trillium/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp#L132-L151

There are a number of other fixes on the `prairie-trillium` branch (https://github.com/rstudio/rstudio/tree/prairie-trillium) that should also be backported.

**Reply by trap000d** [reply]
Thanks, I've just spotted it [v2022.02.2+485](https://github.com/rstudio/rstudio/releases/tag/v2022.02.2%2B485). Indeed it includes the fix.

P.S. I'm pretty sure it wasn't there yesterday, thought it's dated April 20, i.e. 7 days old. Let me guess, you hide releases until finish build and tests :)

**Comment by jjallaire** [comment]
Yes, you are right, the release just came out today (I thought it was already out!).
